### PR TITLE
Add configuration optimization to run SONiC on DPU

### PR DESCRIPTION
**Pull request by oleksandrivantsiv**
#### Why I did it

- The Appliance DPU supports a limited subset of the configuration defined in the [SONiC DASH HLD](https://github.com/sonic-net/SONiC/blob/master/doc/dash/dash-sonic-hld.md).

- Configure SWSS and GMNI to communicate on the Appliance DPU.

- Add a sample configuration for the `appliance` topology.

##### Work item tracking
- Microsoft ADO **(number only)**:

#### How I did it
Disable configuration unsupported by the Appliance DPU based on the device type.

#### How to verify it
Run VNET-to-VNET sonic-mgmt test.

#### Which release branch to backport (provide reason below if selected)

- [ ] 201811
- [ ] 201911
- [ ] 202006
- [ ] 202012
- [ ] 202106
- [ ] 202111
- [ ] 202205
- [ ] 202211
- [ ] 202305

#### Tested branch (Please provide the tested image version)

- [ ] 
- [ ]

### CODE REVIEW

**Review by dgsudharsan — APPROVED** [review]

**Review by Yakiv-Huryk — APPROVED** [review]

File: dockers/docker-orchagent/orchagent.sh [hunk 1]
```diff
@@ -86,6 +86,9 @@ if [[ x"${LOCALHOST_SUBTYPE}" == x"SmartSwitch" ]]; then
     else
         ORCHAGENT_ARGS+=" -q tcp://127.0.0.1:8100"
     fi
+# Enable ZMQ for DPU Appliance
+elif [[ x"${LOCALHOST_SUBTYPE}" == x"Appliance" ]]; then
```
Comment (prsunny) [inline dockers/docker-orchagent/orchagent.sh:90]: I don't see a subtype 'Appliance'. It should be 'SmartSwitch' for both npu and dpu. @Pterosaur

File: dockers/docker-sonic-gnmi/gnmi-native.sh [hunk 2]
```diff
@@ -70,9 +70,9 @@ else
     TELEMETRY_ARGS+=" -v=2"
 fi
 
-# Enable ZMQ for SmartSwitch
+# Enable ZMQ for SmartSwitch or DPU Appliance
 LOCALHOST_SUBTYPE=`sonic-db-cli CONFIG_DB hget "DEVICE_METADATA|localhost" "subtype"`
-if [[ x"${LOCALHOST_SUBTYPE}" == x"SmartSwitch" ]]; then
+if [[ x"${LOCALHOST_SUBTYPE}" == x"SmartSwitch" || x"${LOCALHOST_SUBTYPE}" == x"Appliance" ]]; then
```
Comment (prsunny) [inline dockers/docker-sonic-gnmi/gnmi-native.sh:75]: Same as above, please remove

File: src/sonic-config-engine/config_samples.py [hunk 3]
```diff
@@ -194,13 +195,46 @@ def generate_l2_config(data):
             data['VLAN_MEMBER']['Vlan1000|{}'.format(port)] = {'tagging_mode': 'untagged'}
     return data
 
+def generate_appliance_config(data):
+    data['DEVICE_METADATA']['localhost']['hostname'] = 'sonic'
+    data['DEVICE_METADATA']['localhost']['switch_type'] = 'dpu'
+    data['DEVICE_METADATA']['localhost']['type'] = 'SonicHost'
```
Comment (prsunny) [inline src/sonic-config-engine/config_samples.py:201]: this is SonicDpu

File: dockers/docker-orchagent/orchagent.sh (hunk 1 above)
Comment (oleksandrivantsiv) [reply dockers/docker-orchagent/orchagent.sh]: `Appliance` is not needed. I removed the changes.

File: dockers/docker-sonic-gnmi/gnmi-native.sh (hunk 2 above)
Comment (oleksandrivantsiv) [reply dockers/docker-sonic-gnmi/gnmi-native.sh]: `Appliance` is not needed. I removed the changes.

File: src/sonic-config-engine/config_samples.py (hunk 3 above)
Comment (oleksandrivantsiv) [reply src/sonic-config-engine/config_samples.py]: I removed this file from the PR. I'll raise another PR after we close the question with the default topology name for the DPU.

**Review by prabhataravind — APPROVED** [review]
lgtm

**Review by prsunny — DISMISSED** [review]

**Comment by liat-grozovik** [comment]
@prsunny please help to review the ms_conflict checker and please let us know what is the next step to have this PR included

**Comment by prsunny** [comment]
@liat-grozovik , @oleksandrivantsiv , working on the conflict. Will merge after the conflict is resolved

File: dockers/docker-orchagent/switch.json.j2 [hunk 4]
```diff
@@ -18,13 +18,15 @@
 [
     {
         "SWITCH_TABLE:switch": {
+{% if DEVICE_METADATA.localhost.switch_type != "dpu" %}
```
Comment (prsunny) [inline dockers/docker-orchagent/switch.json.j2:21]: Can you add modify to this -` {% if DEVICE_METADATA.localhost.switch_type and DEVICE_METADATA.localhost.switch_type != "dpu" %}`

Comment (oleksandrivantsiv) [reply dockers/docker-orchagent/switch.json.j2:21]: By default, this configuration should be enabled. We want to disable it only for DPU.
The `switch_type` attribute in DEVICE_METADATA table is optional. When the attribute is missing the assumption is that the type is `switch`. The condition for this should be the following:
`{% if not DEVICE_METADATA.localhost.switch_type or DEVICE_METADATA.localhost.switch_type != "dpu" %}`
@prsunny can you please confirm this?

**Comment by lguohan** [comment]
@prsunny , please faciltate to address ms_conflict test failure.

File: dockers/docker-orchagent/switch.json.j2 (hunk 4 above)
Comment (prsunny) [reply dockers/docker-orchagent/switch.json.j2:21]: @oleksandrivantsiv , I feel thats still not going to achieve the intended functionality with 'or'. One condition will be true when swtich_type is present even with 'dpu' and continues with setting attributes.

**Comment by Pterosaur** [comment]
> @prsunny , please faciltate to address ms_conflict test failure.

@lguohan The ms_conflict is passed.

File: dockers/docker-orchagent/switch.json.j2 (hunk 4 above)
Comment (oleksandrivantsiv) [reply dockers/docker-orchagent/switch.json.j2:21]: The condition is 'if **not** DEVICE_METADATA.localhost.switch_type or DEVICE_METADATA.localhost.switch_type != "dpu"'. 

It should hit in two cases. When the `switch_type` is not present in the table:
image
Or when `switch_type` is not equal to `dpu`
image
And should not hit when the `switch_type` is `dpu`
image

**Review by prsunny — APPROVED** [review]

**Pull request merged**